### ISSUE TITLE
Inline global (exported) functions as well

### ISSUE THREAD

**Issue by MaxGraey** [WebAssembly/binaryen#2921]
It will be great if Binaryen inline function with `usedGlobally` field as well. At least for `options.optimizeLevel >= 3`. Sometimes it even could decrease size if all func arguments are constant and ultimately  folded to one constant.
https://github.com/WebAssembly/binaryen/blob/master/src/passes/Inlining.cpp#L78

Or we could add extra pass option like `alwaysInlineGlobalFunctions`

WDYT?

**Comment by kripken** [comment]
I agree. The current inlining heuristics could be improved a lot - we are too cautious atm.

**Comment by MaxGraey** [comment]
Can I create a PR for this? Or is it better that someone from the core team would do this?

**Comment by kripken** [comment]
In general, for inlining most of the work isn't in writing the PR I think, but in testing on a lot of codebases and seeing if the new heuristics help or not. Running the emscripten test and benchmark suites can help there, but then you often need to investigate the results.

But taking a quick look at this, we already are willing to inline such functions, we end up here:
```cpp
    // more than one use, so we can't eliminate it after inlining,
    // so only worth it if we really care about speed and don't care
    // about size, and if it's lightweight so a good candidate for
    // speeding us up.
    return options.optimizeLevel >= 3 && options.shrinkLevel == 0 &&
           lightweight;
```
So `-O3` will inline functions with global uses, but only if they are lightweight. Changing that would definitely need careful analysis on real-world code, but it's a good idea to investigate that if you or someone else has time.

**Comment by MaxGraey** [comment]
> But taking a quick look at this, we already are willing to inline such functions, we end up here:

I see. makes sense. But `lightweight` will be really depends on runtime and consumer of binaryen. For example after LLVM it could be enough, but for some other backends like after AS or Go it could be better have ability manually skip `lightweight` as option parameter. What about `alwaysInlineGlobals` which `false` by default?

```js
return options.optimizeLevel >= 3 && options.shrinkLevel == 0 &&
           (lightweight || options.alwaysInlineGlobals);
```

**Comment by kripken** [comment]
The name is maybe not clear enough, but I'd be ok with `options.inlining.allowHeavyweight`.

**Comment by MaxGraey** [comment]
implemented in #3032
